File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.py
class Solution:
    def search(self, arr, k):
        n = len(arr)  # size of the array
        left, right = 0, n - 1

        while left <= right:
            mid = (left + right) // 2

            # if mid points to the target
            if arr[mid] == k:
                return True

            if arr[mid] == arr[left] and arr[mid] == arr[right]:
                left += 1
                right -= 1
                continue
            
            if arr[mid] >= arr[left]:
                if k >= arr[left] and k <= arr[mid]:
                    right = mid
                else:
                    left = mid + 1
            else:
                if k >= arr[mid] and k <= arr[right]:
                    left = mid
                else:
                    right = mid - 1

        return False

File: 0081-search-in-rotated-sorted-array-ii/0081-search-in-rotated-sorted-array-ii.java
class Solution {
    public boolean search(int[] nums, int target) {
        int n = nums.length;
        int left = 0, right = n - 1;

        while (left <= right) {
            int mid = (left + right) / 2;
            System.out.println(mid + " " + left + " " + right);
            if (nums[mid] == target) {
                return true;
            }

            if (nums[left] == nums[right] && nums[mid] == nums[right]) {
                left += 1;
                right -= 1;
            }
            else if (nums[left] <= nums[mid]) {
                if (nums[left] <= target && nums[mid] > target) {
                    right = mid - 1;
                } else {
                    left = mid + 1;
                }
            } else {
                if (nums[mid] < target && nums[right] >= target) {
                    left = mid + 1;
                } else {
                    right = mid - 1;
                }
            }
        }

        return false;
    }
}